### ISSUE TITLE
Feature Request: Do not obsolete/deprecate NetMQContext

### ISSUE THREAD

**Issue by MikeFair** [zeromq/netmq#448]
I noticed that manually creating a NetMQContext is being marked as obsolete.  Instead, I propose extending the "new Socket()" constructor methods to take a provided context; e.g "new Socket(m_context) (and perhaps even allow Contexts to be named and some static discovery and management functions to the NetMQContext class).

I like the default simplicity of having a static unnamed context maintained at the library level for a simple Process; but I don't think creating contexts should be removed or marked obsolete.  If for no other reason than porting ZeroMQ code from other languages; but they also provide an extremely useful role for more complex infrastructures and services.   

Contexts create the equivalent of a network switch, VLAN, WiFI SSID, Autonomous System ID, or other organizational/security boundary to group communication areas within a single process.  It creates logical partitions to isolate independent communication scopes within the same process.

For example, using discovery protocols, I would expect every Socket within the same Context to be discoverable to every other Socket within that Context, but I wouldn't expect Sockets across Contexts to be automatically discoverable as part of the same peer Zone.  

The context creates a natural boundary in that sense; establishing isolated discovery zones. If I wanted to create multiple discovery zones within the same Process, manually managing the Context for each Zone is a simple pattern that's easy to understand and makes complete sense from a programming perspective.  One can first discover the "Zones" (aka Contexts; even give them a name) within a process, and then discover all the "Queues" (Sockets) that exist within those Zones.  

I always pictured "Context" as being usable for something equivalent to "domain name"; the same process might be providing its service to multiple "domains" at the same time (like a Web, Video Conference Streaming, or FTP server). The Context is an easy boundary to configure, control, and isolate the collection of Sockets that belong to each separate domain.

The Context can expand to include the role of collecting, storing, managing, and aggregating information about all the Sockets it controls. And methods to request monitoring and inspection at the context level could provide event feeds on the creation/deletion of all Sockets (Queues) within that Context; a "tap" to tee off all of the messages happening on a particular Socket for debugging/inspecting a live running system; and perhaps even a way to manually manage the Sockets themselves (like if the State Machine for a Socket gets out of whack, this method could provide a way to discover the conditions and even allow us a way to manually "fix" the Socket without rebooting the process).

Using Unity as the rendering engine, this idea is one of the things I've been investigating.  Imagine processes hosted at an ISP for you.  The service processes themselves would likely each deal in many different user's domains (The File Server has code that specializes in Disk I/O while the Web Server specializes in rendering Web Pages; but the processes themselves are each handling multiple domains at the same time).  

If these services were created in separate Contexts within each OS Process; one for each Domain they participate in, then that Context object could provide all sorts of useful functions.  

The Unity application on your local laptop could inspect and manage all the services running for you around the globe because it could connect to the isolated context for your service within each process at the various locations.  Enabling Contexts to connect with each other across processes and then machines makes for an internet wide discovery zone for a single administrative domain.

If the Context gets removed, we lose that level of Socket collection, isolation, and management opportunity.

Thanks,
Mike

**Comment by MikeFair** [comment]
I started digging into the code to see if I could create a pull request for adding in a constructor that took a NetMQContext for all the socket classes but it wasn't as obvious as I thought...

I got lost between NetMQContext containing a Ctx object and and NetMQConfig.Context being a Ctx object and the Ctx object being the thing containing all the data for the Context.  It wasn't obvious to me how I could pass a NetMQContext to a NetMQSocket constructor and get access to the necessary Ctx objects.

My plan was that any function signature that require a NetMQContext would use NetMQConfig.Context in its place; the problem was NetMQConfig.Context is a Ctx object and not a NetMQContext.

I did some initial work on making it so the derived Socket classes can be responsible for creating themselves to try and eliminate the switch statements based on ZmqSocketType.

Check out pull request: https://github.com/zeromq/netmq/pull/449

**Comment by somdoron** [comment]
@MikeFair I'm not sure about your catch of context, the context is just a background thread doing the IO of NetMQ. That all it is, also the zeromq community is going away from context, it is just adding unneeded complication. Think of it is operating system doing the IO, you don't know how many threads the OS allocating to manage IO, also you cannot say to the OS I want my IO to be isolated from other IO operations. I think the context is just confusing, when I'm explaining NetMQ to new users and I need to explain the context I just say "it just something you have to create and pass around", that is just wrong.

Regarding managing, isolation, etc, nobody really needs that, if one day someone will come with real problem, we can solve that, but this is kind of might be a problem just complicated the library and add waste.

NetMQ and ZeroMQ is hard libraries to learn and have some learning curve, I think removing the context reducing this learning curve, make the library smaller and less confusing.

**Comment by drewnoakes** [comment]
I agree with the idea that ZeroMQ/NetMQ is hard enough to learn, and anything we can do to simplify it is a good thing. The context gives an incorrect impression of isolation, and getting rid of it does simplify things. Also, some classes were instantiated via methods on the context, others via constructors. Now we have a cleaner API with fewer things to think about.

**Comment by MikeFair** [comment]
I think we're using the idea of Context in two different ways here; so I'm
going to call my Context a ZMQContext, and we'll keep the Context you're
thinking about as Ctx; named for the internal classes already in the
codebase.

And just because many people don't understand what they're looking at
doesn't mean it doesn't make sense or doesn't belong. :)
We can still make it easy/simple for them to use. The biggest challenge
with Zmq I see is that it's got half the concepts of a level 5 OSI model
layer and trying to masquerade as just a layer 4 and people keep trying to
use it like layer 5 but the abstraction layer isn't complete.  I think ZMQ
can become the internet's first meaningful layer 5; And the context is a
menaningful part of that creation.

The ZMQContext is the equivalent to a Network ID, Broadcast domain, or
Logical Scoping.

For example, if there was a method on the ZMQContext object that allowed me
to name the local IP interfaces the ZmqContext should use then
bind(*:5555); would mean something different then it does without the
existence of a ZmqContext at all.

And this ZMQContext thing is something logically different from the Ctx
Context thing.  Though they become siblings to each other (in the same way
a ZmqSocket using a tcp transport is related to a TCP/IP socket -
ZmqContext is could represent an isolated grouping of names and resources).

**Comment by somdoron** [comment]
They are all the same. Currently in ZeroMQ the zmq_context is only providing the IO thread, not isolation, not scoping or netword ID, none of those is supported neither in NetMQ, ZeroMQ or CZMQ. Also zmq_context is kind of obsolete (not officially yet), CZMQ doesn't have context anymore.

But it come down to one question, what problem does not obsoleting context solve today? I understand what it can do in the future, but it doesn't to do it today. Also none of those requests never came as real problems. 

The philosophy of zeromq and NetMQ is solving problems, not providing features, so maybe we can solve the problem you are talking about without context, why not just opening an issue with the problem you are trying to solve and then we can come up with solutions that direct those problems. Maybe the solution will be with context, maybe not.

**Comment by eric-b** [comment]
Hi all,

I wish to add my preference to not make obsolete the legacy NetMqContext:

I think about a third-party library which would use netmq and a main application: if there is a bug (socket not released) in the library, as the main app developer, I would not want to be affected by the third party bug which I cannot control (apart create an issue and be stucked waiting a resolution).

I don't know if Martin Sústrik changed his mind on this topic, but he clearly expressed the case in a [chapter he wrote for an ebook here](http://aosabook.org/en/zeromq.html) (see 24.2 Global State).

I think it is a concrete case.

If you want to make NetMqContext.Create obsolete to notifiy current users the new way of managing sockets, an alternative would be to propose another method to create NetMqContext (perhaps a public constructor or a new method).

Thanks,
Eric

**Comment by somdoron** [comment]
@eric-b regarding martin opinion, I think he changed it, read here:

http://250bpm.com/blog:23

Regarding bugs and etc, i want to change that behavior, you will not be affected by other libraries bugs, see the PR from today:

https://github.com/zeromq/netmq/pull/451

So you are right about the libraries problem, but there are easier solutions to this problem.

**Comment by eric-b** [comment]
Thanks @somdoron, I was not aware of this article of Martin.

So all fine, great library :-)

**Comment by MikeFair** [comment]
That http://250bpm.com/blog:23 post is a good example of wanting to use ZeroMQ as a layer 5 api, they wanted to connext to the right "logger" queue; and they wanted the right transport to be used; and can highlight how the two types of contexts aren't the same; and how with these higher level contexts, the right transport or chain of transports could be used to connect the endpoints.

Here's how:

The example given was a logger and two libraries that should use inproc and not ipc or tcp.  The truth is the application layer should express a desire to connect to the "logger" queue; not necessarily how it wants to connect to the queue.

Imagine for a moment there's this new thing called a Queue, and Sockets can connect to Queues or to other Sockets.

Process B starts up and requests a PushSocket to "Connect" to the "logger" Queue and starts sending data; as there are no listeners at the moment, the messages go in, fill up the Queue, and the Queue settings say to drop packets in FIFO order so as new packets show up, older messages are just dropped.

Process A starts up and also connects to the "logger" Queue with a Pull Socket.  At this point, the ZeroMQ library they are sharing can connect the PushSocket to the PullSocket because it knows they both want to talk to "logger".

Process C shows up, creates a PushSocket and connects to "logger", the library then Connects C's PushSocket with A's PullSocket.  A Queue is a lot like a registered internal ZmqDevice name.

Now for where "Contexts" come into play.  Process D shows up, but Process D is for an entirely different application domain and perhaps even dns domain;  it will connect to the same "logger" because that's what the library does.  However, it needs to identify itself to the logger as part of something different; and this is where the Socket's "Context" can make a difference.  The same reason why the http protocol now takes the domain name as part of the get request, so the service provider can distinguish the "context" of this get message from other messages.

This way there can be a single "logger" service, and it can be shared by multiple entities without confusing anything.

Once you have Contexts, the Contexts can automatically start discovering each other, discovering the service queues they can offer and start setting up efficient Socket links between them; they could be inproc, ipc, tcp, or some other transport we haven't figured out yet.

A Socket can also request to Connect to a Queue in a different Context (think of context name as domain name here).  Because of the Context discovery mechanism, this Socket can be connected to that Queue using the most efficient transport route available and even reroute over other transports if required/configured.

Lastly, Contexts in different places with the same id (again think of domain name as a context id here; though it could be something else) can be more forthcoming about broadcasting the service queues they are covering to give the end users at the application level a powerful method for autodiscovering all the queues available to their context.

So in short, the Context enables isolated namespaces for the collection of Sockets as well as interfaces for transports.  So for example, if I assign eth0 to context A and eth1 to Context B, then Sockets issuing Bind("tcp://*:8888") will have something different occur depending on if the context of the Socket is Context A or Context B.

Thanks
